Source/WebCore/dom/Document.cpp

```diff
@@ -4428,12 +4428,26 @@ String Document::documentURI() const
     );
 }
 
+void Document::setURL(URL&& url)
```

Comment:
These PRs are causing a lot of code duplication without performance numbers. This is very unfortunate. Several ways to avoid that:
1. Keep only the move version and port all call sites to it
2. Update the parameter to pass the URL by value. Then the internal implementation would do a WTFMove(). And the call sites can either WTFMove() the parameter in or not, with a single implementation.

cc @geoffreygaren @darinadler @weinig for opinions.

Reply:
I strongly agree with your overall point, @cdumez. I also agree that either of your suggestions would be a good approach. Your suggestion (2) is more elegant at call sites, and your suggestion (1) is less elegant but slightly decreases the chance that we’ll accidentally keep reference count churn in cases where we can avoid it. I think I prefer (2).

Reply:
Yeah, in cases with mixed callers (T& and T&&) I like (2). You can always switch to (1) eventually if all callers become T&&.

Reply:
@cdumez So if I modify this code so that instead of a (T&&) and a (const T&) implementation, we just have a single (T) implementation, moves will happen in all cases where we remember to WTFMove() the argument, or it the result of a RVO call?

Reply:
Yes.

Notably, pass by value is only a good approach here because you are always going to either (a) move the URL, or (b) copy the URL. Normally you would only want to use pass by value with primitive types or very small objects, but `Document::setURL` is going to copy that URL one way or the other, so it really makes sense to just use pass by value here.

Reply:
> moves will happen in all cases where we remember to WTFMove() the argument
Right. There will be zero copy if the call sites uses WTFMove(). There will be a single copy when they don't.

Reply:
Note there are [core guidelines](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#fcall-parameter-passing) covering this. I ought to read those....

Reply:
A more difficult case than this one, which presents a small design challenge, is a function that often examines its argument and only sometimes moves the value of argument in. In that case a `const T&` can be more efficient than `T` for the case where the argument value is not moved in, so it's trickier to decide which solution is best.

Reply:
Another tricky case is when T is a large type (like FontDescription) rather than a pointer wrapper (like URL). In that case, I’m not sure that a WTFMove inside the callee is enough to avoid copying memory.

Reply:
> Another tricky case is when T is a large type (like FontDescription) rather than a pointer wrapper (like URL). In that case, I’m not sure that a WTFMove inside the callee is enough to avoid copying memory.

Excellent point. Here’s the shorthand way I think about that: If it’s not inlined or link-time optimized there are 2 moves and 0 copies in this idiom. Thus, if the move is expensive enough, might need to use `T&&` so it’s only 1 move and 0 copies.

Reply:
URL is actually not just a pointer wrapper: [it's filled with integers](https://github.com/WebKit/WebKit/blob/fa3ef110e7667e0299e44ed2959ac1e4890f4154/Source/WTF/wtf/URL.h#L272).

Reply:
> URL is actually not just a pointer wrapper: [it's filled with integers](https://github.com/WebKit/WebKit/blob/fa3ef110e7667e0299e44ed2959ac1e4890f4154/Source/WTF/wtf/URL.h#L272).

True, URL is actually quite large. The good news is that I believe Brent is working on updating his PR to use `T&&` only.